include/beast/core/flat_streambuf.hpp

```diff
@@ -0,0 +1,297 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_FLAT_STREAMBUF_HPP
+#define BEAST_FLAT_STREAMBUF_HPP
+
+#include <beast/config.hpp>
+#include <beast/core/detail/empty_base_optimization.hpp>
+#include <boost/asio/buffer.hpp>
+#include <limits>
+#include <memory>
+
+namespace beast {
+
+/** A linear dynamic buffer.
+
+    Objects of this type meet the requirements of
+    @b `DynamicBuffer` and offer an additional invariant:
+    Buffer sequences returned by @ref data and @ref prepare
+    will always be of length one.
+
+    @note This class is designed for use with algorithms that
+    take dynamic buffers as parameters, and are optimized
+    for the case where the input sequence or output sequence
+    is stored in a single contiguous buffer.
+*/
+template<class Allocator>
+class basic_flat_streambuf
+#if ! BEAST_DOXYGEN
+    : private detail::empty_base_optimization<
+        typename std::allocator_traits<Allocator>::
+            template rebind_alloc<char>>
+#endif
+{
+public:
+#if BEAST_DOXYGEN
+    /// The type of allocator used.
+    using allocator_type = Allocator;
+#else
+    using allocator_type = typename
+        std::allocator_traits<Allocator>::
+            template rebind_alloc<char>;
+#endif
+
+private:
+    template<class OtherAlloc>
+    friend class basic_flat_streambuf;
+
+    using alloc_traits =
+        std::allocator_traits<allocator_type>;
+
+    static
+    inline
+    std::size_t
+    dist(char const* first, char const* last)
+    {
+        return static_cast<std::size_t>(last - first);
```

Comment:
Its not necessary, since we know these are pointer types. It would unnecessarily drag in `<algorithm>` only to compile to the same code.